src/ansible_navigator/ui_framework/utils.py

```diff
@@ -3,28 +3,48 @@
 import functools
 import re
 
+from dataclasses import is_dataclass
 from math import floor
+from typing import Any
+from typing import Dict
 from typing import List
+from typing import Union
 
 
-def convert_percentage(data_dict: dict, keys: List, progress_bar_width: int) -> None:
-    """convert a string % to a little progress bar
-    not recursive
-    80% = 80%|XXXXXXXX  |
+def convert_percentage(
+    content: Union[Dict[str, Any], Any],
+    columns: List[str],
+    progress_bar_width: int,
+) -> None:
+    """Convert an attribute or value to a progress bar formatted for the TUI in place.
 
-    :param data_dict: The dictionary to update
-    :param keys: The keys to convert in each dictionary
-    :param progress_bar_width: The width of the progress bar
+    :param content: The content for which the progress bar will be generated
+    :param columns: The menu columns, only make progress bars for columns
+    :param progress_bar_width: The target width of the progress bar
     """
-    for key in keys:
-        value = data_dict.get(key)
+    for column in columns:
+        value = content.get(column)
         if value and is_percent(str(value)):
-            if value == "100%":
-                data_dict[key] = "COMPLETE".center(progress_bar_width, " ")
-            else:
-                chars_in_progress_bar = floor(progress_bar_width / 100 * int(value[0:-1]))
-                data_dict["_" + key] = value
-                data_dict[key] = ("\u2587" * chars_in_progress_bar).ljust(progress_bar_width)
+            new_value = _string_to_progress(value, progress_bar_width)
+            if isinstance(content, dict):
+                content["_" + column] = value
+                content[column] = new_value
+            elif is_dataclass(content):
```

Comment:
What happens if neither of these cases is true, since the type currently says that can happen?

Reply:
The Any there is just a place holder for ContentBase in https://github.com/ansible/ansible-navigator/pull/1014

At that point, mypy should protect it.  If something did slip through one of a few things would happen:

1) There would be a traceback because the object didn't have a `get` function
2) There would be traceback because the object didn't have a `_` attribute corresponing to the column name that had a %
3) The object would not be modified if it did not have any `%` strings, did have a get.